crates/bevy_diagnostic/src/diagnostic.rs

```diff
@@ -129,40 +129,97 @@ impl Diagnostic {
     }
 }
 
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+pub enum DiagnosticState {
+    Enabled,
+    Disabled,
+}
+
 /// A collection of [Diagnostic]s
 #[derive(Debug, Default)]
 pub struct Diagnostics {
     // This uses a [`StableHashMap`] to ensure that the iteration order is deterministic between
     // runs when all diagnostics are inserted in the same order.
-    diagnostics: StableHashMap<DiagnosticId, Diagnostic>,
+    diagnostics: StableHashMap<(DiagnosticId, DiagnosticState), Diagnostic>,
```

Comment:
I fixed the `add`, `disable`, `add` scenario by removing any potential `Disabled` diagnostic before re-adding it...

Yup, it's to avoid the cost of the extra check and... the fact that the code became not pretty by adding two layers of `if`
```rust
    // with the state in the key
    pub fn get(&self, id: DiagnosticId) -> Option<&Diagnostic> {
        self.diagnostics.get(&(id, DiagnosticState::Enabled))
    }

    // with the state as an property
    pub fn get(&self, id: DiagnosticId) -> Option<&Diagnostic> {
        if let Some(diagnostic) = self.diagnostics.get(&id) {
            if diagnostic.state == DiagnosticState::Enabled {
                return Some(diagnostic);
            }
        }
        None
    }
```

As I write this, it just came to mind that it would have work with
```rust
    pub fn get(&self, id: DiagnosticId) -> Option<&Diagnostic> {
        self.diagnostics
            .get(&id)
            .filter(|diagnostic| diagnostic == DiagnosticState::Enabled)
    }
```

Still I find the state as key to be prettier, but I'm open to change

Reply:
That filter could probably also be `Diagnostic::is_enabled`

I find that this is a messy implementation, especially given that it makes the `state` method `O(n)`. If I'd found this during one of my archaeology sessions, I'd add it to my refactorings list.

Now maybe the performance effects are worthwhile, but I'm exceedingly skeptical.

Reply:
you convinced me 👍